### PR TITLE
Canonicalize provider version during default provider lookup

### LINKED ISSUE #16108: Pulumi loads multiple default providers for same package
### What happened?

I wrote a NodeJS program that uses the default kubernetes provider, and found that two providers were instantiated in stack state. I expected one.

### Example

```ts
import * as k8s from "@pulumi/kubernetes";

new k8s.yaml.v2.ConfigGroup("example", {
    objs: [
        {
            kind: "ConfigMap",
            apiVersion: "v1",
            metadata: {
                name: "example"
            }
        }
    ]
});
```

Stack state (excerpt):

```jsonl
{
    "urn": "urn:pulumi:dev::issue-1939-nodejs::pulumi:providers:kubernetes::default_4_10_0",
    "custom": true,
    "id": "951e952b-98f5-471b-9fc8-4331ee69a824",
    "type": "pulumi:providers:kubernetes",
    "inputs": {
        "version": "4.10.0"
    },
    "outputs": {
        "version": "4.10.0"
    },
    "created": "2024-05-02T21:23:46.511976Z",
    "modified": "2024-05-02T21:23:46.511976Z"
}
{
    "urn": "urn:pulumi:dev::issue-1939-nodejs::pulumi:providers:kubernetes::default",
    "custom": true,
    "id": "5a493c4c-bd7c-4727-b54c-8ed025c5d844",
    "type": "pulumi:providers:kubernetes",
    "inputs": {
        "version": "4.10.0"
    },
    "outputs": {
        "version": "4.10.0"
    },
    "created": "2024-05-02T21:23:47.163822Z",
    "modified": "2024-05-02T21:23:47.163822Z"
}
{
    "urn": "urn:pulumi:dev::issue-1939-nodejs::kubernetes:yaml/v2:ConfigGroup$kubernetes:core/v1:ConfigMap::example:eron/example",
    "custom": true,
    "id": "eron/example",
    "type": "kubernetes:core/v1:ConfigMap",
    "parent": "urn:pulumi:dev::issue-1939-nodejs::kubernetes:yaml/v2:ConfigGroup::example",
    "provider": "urn:pulumi:dev::issue-1939-nodejs::pulumi:providers:kubernetes::default::5a493c4c-bd7c-4727-b54c-8ed025c5d844",
},
```

### Output of `pulumi about`

```
❯ p about
CLI          
Version      3.113.3
Go Version   go1.22.2
Go Compiler  gc

Plugins
KIND      NAME        VERSION
resource  kubernetes  4.10.0
language  nodejs      unknown

Host     
OS       darwin
Version  14.4.1
Arch     arm64

This project is written in nodejs: executable='/opt/homebrew/bin/node' version='v21.6.2'

Current Stack: eron-pulumi-corp/issue-1939-nodejs/dev

TYPE                            URN
pulumi:pulumi:Stack             urn:pulumi:dev::issue-1939-nodejs::pulumi:pulumi:Stack::issue-1939-nodejs-dev
pulumi:providers:kubernetes     urn:pulumi:dev::issue-1939-nodejs::pulumi:providers:kubernetes::default_4_10_0
kubernetes:yaml/v2:ConfigGroup  urn:pulumi:dev::issue-1939-nodejs::kubernetes:yaml/v2:ConfigGroup::example
pulumi:providers:kubernetes     urn:pulumi:dev::issue-1939-nodejs::pulumi:providers:kubernetes::default
kubernetes:core/v1:ConfigMap    urn:pulumi:dev::issue-1939-nodejs::kubernetes:yaml/v2:ConfigGroup$kubernetes:core/v1:ConfigMap::example:eron/example
pulumi:providers:pulumi         urn:pulumi:dev::issue-1939-nodejs::pulumi:providers:pulumi::default


Found no pending operations associated with dev

Dependencies:
NAME                VERSION
@pulumi/kubernetes  4.10.0
@pulumi/pulumi      3.114.0
@types/node         18.19.31
typescript          5.4.5

```

### Additional context

An important detail is that the `ConfigGroup` resource is an MLC based on the Go SDK. Some of the calls to `RegisterResource` include a `version` option, some do not, based on which SDK makes the call. The default provider lookup logic seems to create two providers as a result.

In the above example, the Go SDK created `urn:pulumi:dev::issue-1939-nodejs::pulumi:providers:kubernetes::default` because it didn't supply a `version` option.

### Logs

First snippet, showing registration of the versioned provider:
```
I0502 09:57:35.006440   99098 eventsink.go:59] resource registration prepared: ty=kubernetes:yaml/v2:ConfigFile, name=calico-operator
I0502 09:57:35.008561   99098 rpc.go:292] Unmarshaling property for RPC[ResourceMonitor.RegisterResource(kubernetes:yaml/v2:ConfigFile,calico-operator)]
I0502 09:57:35.008580   99098 source_eval.go:396] handling default provider request for package kubernetes-4.11.0
I0502 09:57:35.008867   99098 source_eval.go:452] checking if providers.ProviderRequest{version:(*semver.Version)(0x14001f7dae0), pkg:"kubernetes", pluginDownloadURL:"", pluginChecksums:map[string][]uint8(nil)} should be denied
I0502 09:57:35.008889   99098 source_eval.go:326] newRegisterDefaultProviderEvent(kubernetes-4.11.0): using version 4.11.0 from request
I0502 09:57:35.008893   99098 source_eval.go:345] newRegisterDefaultProviderEvent(kubernetes-4.11.0): no pluginDownloadURL specified, falling back to default pluginDownloadURL
I0502 09:57:35.008897   99098 source_eval.go:353] newRegisterDefaultProviderEvent(kubernetes-4.11.0): default pluginDownloadURL miss, sending empty string to engine
I0502 09:57:35.008899   99098 source_eval.go:363] newRegisterDefaultProviderEvent(kubernetes-4.11.0): no pluginChecksums specified, falling back to default pluginChecksums
I0502 09:57:35.008902   99098 source_eval.go:371] newRegisterDefaultProviderEvent(kubernetes-4.11.0): default pluginChecksums miss, sending empty map to engine
I0502 09:57:35.008931   99098 source_eval.go:429] waiting for default provider for package kubernetes-4.11.0
I0502 09:57:35.008934   99098 source_eval.go:191] EvalSourceIterator produced a registration: t=pulumi:providers:kubernetes,name=default_4_11_0,#props=1
I0502 09:57:35.394220   99098 source_eval.go:438] registered default provider for package kubernetes-4.11.0: urn:pulumi:dev::issue-1939::pulumi:providers:kubernetes::default_4_11_0
```

Second, showing registration of unversioned provider:
```
I0502 09:57:36.042411   99098 rpc.go:292] Unmarshaling property for RPC[ResourceMonitor.RegisterResource(kubernetes:core/v1:Namespace,calico-operator:tigera-operator)]
I0502 09:57:36.042450   99098 source_eval.go:762] parseProviderRequest(kubernetes): semver version is the empty string
I0502 09:57:36.042456   99098 source_eval.go:396] handling default provider request for package kubernetes
I0502 09:57:36.042464   99098 source_eval.go:452] checking if providers.ProviderRequest{version:(*semver.Version)(nil), pkg:"kubernetes", pluginDownloadURL:"", pluginChecksums:map[string][]uint8(nil)} should be denied
I0502 09:57:36.042479   99098 source_eval.go:329] newRegisterDefaultProviderEvent(kubernetes): no version specified, falling back to default version
I0502 09:57:36.042482   99098 source_eval.go:332] newRegisterDefaultProviderEvent(kubernetes): default version hit on version 4.11.0
I0502 09:57:36.042484   99098 source_eval.go:345] newRegisterDefaultProviderEvent(kubernetes): no pluginDownloadURL specified, falling back to default pluginDownloadURL
I0502 09:57:36.042486   99098 source_eval.go:353] newRegisterDefaultProviderEvent(kubernetes): default pluginDownloadURL miss, sending empty string to engine
I0502 09:57:36.042488   99098 source_eval.go:363] newRegisterDefaultProviderEvent(kubernetes): no pluginChecksums specified, falling back to default pluginChecksums
I0502 09:57:36.042490   99098 source_eval.go:371] newRegisterDefaultProviderEvent(kubernetes): default pluginChecksums miss, sending empty map to engine
I0502 09:57:36.042502   99098 source_eval.go:429] waiting for default provider for package kubernetes
I0502 09:57:36.042504   99098 source_eval.go:191] EvalSourceIterator produced a registration: t=pulumi:providers:kubernetes,name=default,#props=1
I0502 09:57:36.093922   99098 source_eval.go:438] registered default provider for package kubernetes: urn:pulumi:dev::issue-1939::pulumi:providers:kubernetes::default

```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by EronWright**
# Description

This PR addresses a problem that leads to two default providers for a given package+version. The problem occurs when given a combination of unversioned and versioned provider requests, as can occur in multi-language programs.

The fix is as follows: given an unversioned provider request, apply the default version before looking into the provider cache.

Fixes #16108 

## Example
Given the example program from #16108 plus this fix, the stack state has one provider as expected:

```jsonl
{
    "urn": "urn:pulumi:dev::issue-1939-nodejs::pulumi:providers:kubernetes::default_4_10_0",
    "custom": true,
    "id": "c19b65b0-dd5a-432f-8c75-d7544de782ec",
    "type": "pulumi:providers:kubernetes",
    "inputs": {
        "version": "4.10.0"
    },
    "outputs": {
        "version": "4.10.0"
    },
    "created": "2024-05-02T21:22:09.72166Z",
    "modified": "2024-05-02T21:22:09.72166Z"
}
{
    "urn": "urn:pulumi:dev::issue-1939-nodejs::kubernetes:yaml/v2:ConfigGroup$kubernetes:core/v1:ConfigMap::example:eron/example",
    "custom": true,
    "id": "eron/example",
    "type": "kubernetes:core/v1:ConfigMap",
    "parent": "urn:pulumi:dev::issue-1939-nodejs::kubernetes:yaml/v2:ConfigGroup::example",
    "provider": "urn:pulumi:dev::issue-1939-nodejs::pulumi:providers:kubernetes::default_4_10_0::c19b65b0-dd5a-432f-8c75-d7544de782ec",
    "propertyDependencies": {
        "apiVersion": [],
        "kind": [],
        "metadata": []
    },
}
```

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: pkg/resource/deploy/source_eval.go
Comment (EronWright) [inline pkg/resource/deploy/source_eval.go:1]: Note that the "normalization" logic will execute more often, i.e. for cache hits and for misses.

File: pkg/resource/deploy/source_eval_test.go [hunk 1]
```diff
@@ -2042,47 +2042,46 @@ func (c *configSourceMock) GetPackageConfig(pkg tokens.Package) (resource.Proper
...
+	t.Run("normalizeProviderRequest", func(t *testing.T) {
 		t.Parallel()
-		t.Run("error in GetPackageConfig()", func(t *testing.T) {
+		t.Run("use defaultProvider", func(t *testing.T) {
 			t.Parallel()
-			expectedErr := errors.New("expected error")
+			v1 := semver.MustParse("0.1.0")
 			d := &defaultProviders{
+				defaultProviderInfo: map[tokens.Package]workspace.PluginSpec{
+					tokens.Package(""): {
```
Comment (Frassle) [inline pkg/resource/deploy/source_eval_test.go:2052]: bit confusing this being an empty string, can we change this to "pkg" and fill in the `providers.ProviderRequest` below to say "pkg" as well.

Comment (EronWright) [reply pkg/resource/deploy/source_eval_test.go:2052]: That's what the existing test code did but I'll change it.

Comment (Frassle) [reply pkg/resource/deploy/source_eval_test.go]: Thanks

**Review by Frassle — APPROVED** [review]

**Pull request merged**